### PR TITLE
Avoid oversized storage for reduced mse_loss / smooth_l1_loss

### LINKED ISSUE #185647: Unexpected storage from MSE loss output
### 🐛 Describe the bug

MSE loss output takes up more storage than expected even though its not a view.

```python
import torch
import torch.nn.functional as F

x = torch.rand(3, 256, 256)
recon = torch.rand(3, 256, 256)

l2_loss = F.mse_loss(x, recon, reduction='mean')
print(l2_loss)
print(l2_loss.untyped_storage().nbytes())
print(l2_loss._base)
```
Output:
```
tensor(0.1674)
786432
None
```

This behavior doesn't occur with other reductions and some other losses (e.g., F.l1_loss).

### Versions

PyTorch version: 2.9.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.6.1 (arm64)
GCC version: Could not collect
Clang version: 17.0.0 (clang-1700.0.13.5)
CMake version: Could not collect
Libc version: N/A

Python version: 3.11.13 (main, Jun  3 2025, 18:38:25) [Clang 17.0.0 (clang-1700.0.13.3)] (64-bit runtime)
Python platform: macOS-15.6.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Apple M3

Versions of relevant libraries:
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.2.5
[pip3] torch==2.9.0
[pip3] torch-fidelity==0.3.0
[pip3] torchaudio==2.9.0
[pip3] torchcodec==0.9.1
[pip3] torchvision==0.24.0
[pip3] torchviz==0.0.3
[conda] Could not collect

### PR DESCRIPTION
**Pull request by ArsalanShakil**
Fixes #185647

### Summary
`F.mse_loss(x, y, reduction='mean')` returned a scalar whose underlying storage was sized for the full elementwise result (e.g. 786432 bytes for a 3x256x256 input) rather than a single element. `smooth_l1_loss` had the same problem; `l1_loss` and `huber_loss` were already fine.

### Root cause
Both ops are structured. Their meta functions call `build_borrowing_binary_op(maybe_get_output(), ...)`, which sets the output to the full broadcasted elementwise shape and allocates storage for it, then for Mean/Sum reduction call `maybe_get_output().resize_({})`. `resize_` to a smaller size shrinks the logical shape but does not free the storage, so the scalar result kept the full-size allocation.

### Fix
Replace `resize_({})` with `set_output_raw_strided(0, {}, {}, maybe_get_output().options())`. In the structured framework this creates a fresh scalar output (functional variant) instead of reusing the oversized one, reusing the dtype/device that `build_borrowing_binary_op` already computed (so type promotion is preserved). The Mean/Sum impl builds its own elementwise iterator and writes the reduced value via `mean_out`/`sum_out`, so it does not depend on the discarded full-size output.

### Test Plan
```
python test/test_nn.py -k test_reduced_loss_storage_size
python test/test_nn.py -k test_mse_loss_mixed_dtype_grad
python test/test_nn.py -k test_weighted_mse_loss
python test/test_nn.py -k test_mse_loss_size_warning
lintrunner -a aten/src/ATen/native/Loss.cpp test/test_nn.py
```
All pass; reduced `mse_loss`/`smooth_l1_loss` now report single-element storage, `reduction='none'` is unchanged, and values, type promotion, autograd and `torch.compile` all behave correctly.

### CODE REVIEW

**Comment by zou3519** [comment]
@claude review this please and put the review inside a GitHub collapsible section

**Review by zou3519 — CHANGES_REQUESTED** [review]
claude review sounded reasonable to me, but please let me know if you disagree

**Comment by ArsalanShakil** [comment]
I will work on the requested changes 👍

**Comment by ArsalanShakil** [comment]
@zou3519 The use-after-free diagnosis is correct — the `set_output_raw_strided(0, {}, {}, ...)` call re-entered the codegen functional override (`outputs_[0] = create_out(...)`), freeing the TensorImpl that `build_borrowing_binary_op` had borrowed into `operands_[0]`, which `TensorIteratorBase::set_output_raw_strided` then dereferenced. Reworked:

- **Fix (Loss.cpp):** replaced the `set_output_raw_strided` call in both `mse_loss` and `smooth_l1_loss` metas with an in-place mutation of the borrowed output — `out.set_()` (detach the full-size storage) then `out.resize_({})`. This keeps the same TensorImpl the iterator borrows (the original pre-bug control flow used `maybe_get_output().resize_({})`; this only adds the `set_()` to actually drop the storage), so no `outputs_[idx]` reassignment and no dangling borrow.
- **Memory-profiler golden:** regenerated `test_memory_timeline` — the elementwise buffer is now a freed `TEMPORARY` (create+destroy) instead of a persisting `ACTIVATION`.
- **Test:** added a `reduction='none'` assertion that the full elementwise storage is unchanged.

Verified locally (arm64, non-ASAN): `test_reduced_loss_storage_size`, `test_modules.py -k MSELoss/SmoothL1Loss`, `test_dtypes_nn_functional_{mse,smooth_l1}_loss_cpu`, and `test_memory_timeline` all pass; the no-batch-dim and out= paths return single-element storage.

The UAF only manifests under ASAN, which I can't run on this platform (macOS/arm64) — the rework avoids it by construction (it never reassigns the borrowed output operand), and I'm relying on the clang18-ASAN CI here to confirm. The `too many values to unpack` py3.14 failures are unrelated (present on the merge base).

**Comment by ArsalanShakil** [comment]
@zou3519 friendly ping on this one — the requested changes were addressed back in June, and I've now rebased onto current `main` (`35769f8`) so there's a fresh CI signal to review against.

Recap of the rework, since it's been a while:

Your use-after-free diagnosis was right. The original `set_output_raw_strided(0, {}, {}, ...)` call re-entered the codegen functional override (`outputs_[0] = create_out(...)`), which freed the `TensorImpl` that `build_borrowing_binary_op` had borrowed into `operands_[0]` — and `TensorIteratorBase::set_output_raw_strided` then dereferenced it.

The current fix mutates the borrowed output in place instead: `out.set_()` to detach the oversized storage, then `out.resize_({})` to size it to a scalar. No reassignment of `outputs_[idx]`, so the iterator's borrow stays valid by construction. `test_memory_timeline`'s golden is regenerated accordingly — the elementwise buffer is now a freed `TEMPORARY` rather than a persisting `ACTIVATION`.

Would appreciate a re-review when you have a moment.
